client/src/procedures.rs

```diff
@@ -0,0 +1,440 @@
+        String::from_utf8(value.0)
+    }
+}
+
+impl<const N: usize> From<[u8; N]> for ProcedureIo {
+    fn from(a: [u8; N]) -> Self {
+        ProcedureIo(a.into())
+    }
+}
+
+impl<const N: usize> TryFrom<ProcedureIo> for [u8; N] {
+    type Error = <[u8; N] as TryFrom<Vec<u8>>>::Error;
+
+    fn try_from(value: ProcedureIo) -> Result<Self, Self::Error> {
+        value.0.try_into()
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct CollectedOutput {
+    output: HashMap<OutputKey, ProcedureIo>,
+}
+
+impl CollectedOutput {
+    pub fn take<T>(&mut self, key: &OutputKey) -> Option<T>
+    where
+        ProcedureIo: Into<T>,
+    {
+        self.output.remove(key).map(|v| v.into())
+    }
+
+    pub fn try_take<T>(&mut self, key: &OutputKey) -> Option<Result<T, anyhow::Error>>
+    where
+        ProcedureIo: TryInto<T>,
+    {
+        let o = self.output.remove(key)?;
+        Some(o.try_into().map_err(|_| anyhow::anyhow!("Invalid format.")))
+    }
+}
+
+impl IntoIterator for CollectedOutput {
+    type IntoIter = <HashMap<OutputKey, ProcedureIo> as IntoIterator>::IntoIter;
+    type Item = <HashMap<OutputKey, ProcedureIo> as IntoIterator>::Item;
+    fn into_iter(self) -> Self::IntoIter {
+        self.output.into_iter()
+    }
+}
+
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct OutputKey(String);
+
+impl OutputKey {
+    pub fn new<K: ToString>(key: K) -> Self {
+        OutputKey(key.to_string())
+    }
+
+    pub fn random() -> Self {
+        OutputKey(rand::random::<char>().into())
+    }
+}
+
+// ==========================
+// Traits
+// ==========================
+
+pub trait ProcedureStep {
+    fn execute<R: Runner>(self, runner: &mut R, state: &mut State) -> Result<(), anyhow::Error>;
+
+    fn then<P>(self, next: P) -> Procedure
+    where
+        Self: Into<Procedure>,
+        P: Into<Procedure>,
+    {
+        let mut procedure = self.into();
+        procedure.inner.extend(next.into().inner);
+        procedure
+    }
+}
+
+pub trait Runner {
+    fn get_guard<F, T>(&mut self, location0: &Location, f: F) -> Result<T, anyhow::Error>
+    where
+        F: FnOnce(GuardedVec<u8>) -> Result<T, engine::Error>;
+
+    fn exec_proc<F, T>(
+        &mut self,
+        location0: &Location,
+        location1: &Location,
+        hint: RecordHint,
+        f: F,
+    ) -> Result<T, anyhow::Error>
+    where
+        F: FnOnce(GuardedVec<u8>) -> Result<Products<T>, engine::Error>;
+
+    fn write_to_vault(&mut self, location1: &Location, hint: RecordHint, value: Vec<u8>) -> Result<(), anyhow::Error>;
+
+    fn revoke_data(&mut self, location: &Location) -> Result<(), anyhow::Error>;
+
+    fn garbage_collect(&mut self, vault_id: VaultId) -> Result<(), anyhow::Error>;
+}
+
+// ==========================
+//  Traits for the `Procedure` derive-macro
+// ==========================
+
+pub struct Products<T> {
+    pub secret: Vec<u8>,
+    pub output: T,
+}
+
+trait Parse {
```

Comment:
Better use nouns for traits or types, but that's minimal

Reply:
I don't think there is a fixed convention for this; the traits of `std` are not really consistent on this, e.g.: 

- `Borrow`, `Debug`: Verbs
- `Iterator`: Noun
- `Sized`: Adjective